lib/client/src/swap/sanctum.rs

```diff
@@ -0,0 +1,334 @@
+            .timeout(self.timeout_duration)
+            .send()
+            .await
+            .context("quote request to sanctum")?;
+        let quote: QuoteResponse =
+            util::http_error_handling(response).await.with_context(|| {
+                format!("error requesting sanctum route between {input_mint} and {output_mint} (using url: {})", config.sanctum_url)
+            })?;
+
+        Ok(quote)
+    }
+
+    /// Find the instructions and account lookup tables for a sanctum swap through mango
+    pub async fn prepare_swap_transaction(
+        &self,
+        input_mint: Pubkey,
+        output_mint: Pubkey,
+        max_slippage_bps: u64,
+        quote: &QuoteResponse,
+    ) -> anyhow::Result<TransactionBuilder> {
+        tracing::info!("swapping using sanctum");
+
+        let source_token = self.mango_client.context.token_by_mint(&input_mint)?;
+        let target_token = self.mango_client.context.token_by_mint(&output_mint)?;
+
+        let bank_ams = [source_token.first_bank(), target_token.first_bank()]
+            .into_iter()
+            .map(util::to_writable_account_meta)
+            .collect::<Vec<_>>();
+
+        let vault_ams = [source_token.first_vault(), target_token.first_vault()]
+            .into_iter()
+            .map(util::to_writable_account_meta)
+            .collect::<Vec<_>>();
+
+        let owner = self.mango_client.owner();
+        let account = &self.mango_client.mango_account().await?;
+
+        let token_ams = [source_token.mint, target_token.mint]
+            .into_iter()
+            .map(|mint| {
+                util::to_writable_account_meta(
+                    anchor_spl::associated_token::get_associated_token_address(&owner, &mint),
+                )
+            })
+            .collect::<Vec<_>>();
+
+        let source_loan = quote
+            .in_amount
+            .as_ref()
+            .map(|v| u64::from_str(v).unwrap())
+            .unwrap_or(0);
+        let loan_amounts = vec![source_loan, 0u64];
+        let num_loans: u8 = loan_amounts.len().try_into().unwrap();
+
+        // This relies on the fact that health account banks will be identical to the first_bank above!
+        let (health_ams, _health_cu) = self
+            .mango_client
+            .derive_health_check_remaining_account_metas(
+                account,
+                vec![source_token.token_index, target_token.token_index],
+                vec![source_token.token_index, target_token.token_index],
+                vec![],
+            )
+            .await
+            .context("building health accounts")?;
+
+        let config = self.mango_client.client.config();
+
+        let in_amount = quote
+            .in_amount
+            .clone()
+            .expect("sanctum require a in amount");
+        let in_amount_u64 = in_amount.parse::<u64>()?;
+        let out_amount =
+            ((in_amount_u64 as f64) * (1.0 - (max_slippage_bps as f64) / 10_000.0)).ceil() as u64;
```

Comment:
the out amount depending on the in amount looks odd to me. If you convert 1 SOL into jitoSOL you'll only get 0.9 jitoSOL or sth

it's actually unclear what we want here. do we want slippage from quoted amount or slippage from oracle? I think jupiter currently does slippage from quoted amount, so this could be the same (even if it's not actually what rebalancers care about?)

in that case, this should apply the slippage to the quote.out_amount, I think?